keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java

```diff
@@ -24,76 +24,116 @@
  */
 public class INDArrayShim {
 
-    public static void muli(INDArray left, INDArray right, INDArray result) {
+    public static INDArray muli(INDArray left, INDArray right) {
         if (Arrays.equals(left.shape(), right.shape())) {
-            left.muli(right);
+            return left.muli(right);
         } else {
-            broadcastMultiply(left, right, result);
+            return broadcastMultiply(left, right);
         }
     }
 
-    private static void broadcastMultiply(INDArray a, INDArray b, INDArray result) {
-        int[] broadcastDimensions = Shape.getBroadcastDimensions(a.shape(), b.shape());
-        execBroadcast(a, b,
-            new BroadcastMulOp(a, b, result, broadcastDimensions)
+    private static INDArray broadcastMultiply(INDArray a, INDArray b) {
+        int[] broadcastDimensions = getBroadcastDimensions(a.shape(), b.shape());
```

Comment:
Should we do the 'if' first?  This seems like dead code if we're just going to make the recursive call.

Reply:
yes great shout